jdaviz/configs/cubeviz/plugins/spectral_extraction/spectral_extraction.py

```diff
@@ -0,0 +1,132 @@
+
+from jdaviz.core.events import SnackbarMessage
+from jdaviz.core.registries import tray_registry
+from jdaviz.core.template_mixin import (PluginTemplateMixin,
+                                        SelectPluginComponent,
+                                        SpatialSubsetSelectMixin,
+                                        AddResultsMixin)
+from jdaviz.core.user_api import PluginUserApi
+
+__all__ = ['SpectralExtraction']
+
+
+@tray_registry(
+    'cubeviz-spectral-extraction', label="Spectral Extraction", viewer_requirements='spectrum'
+)
+class SpectralExtraction(PluginTemplateMixin, SpatialSubsetSelectMixin, AddResultsMixin):
+    """
+    See the :ref:`Spectral Extraction Plugin Documentation <spex>` for more details.
+
+    Only the following attributes and methods are available through the
+    :ref:`public plugin API <plugin-apis>`:
+
+    * :meth:`~jdaviz.core.template_mixin.PluginTemplateMixin.show`
+    * :meth:`~jdaviz.core.template_mixin.PluginTemplateMixin.open_in_tray`
+    * ``spatial_subset`` (:class:`~jdaviz.core.template_mixin.SubsetSelect`):
+      Subset to use for the spectral extraction, or ``No Subset``.
+    * ``add_results`` (:class:`~jdaviz.core.template_mixin.AddResults`)
+    * :meth:`collapse`
+    """
+    template_file = __file__, "spectral_extraction.vue"
+    function_items = List().tag(sync=True)
+    function_selected = Unicode('Sum').tag(sync=True)
+
+    def __init__(self, *args, **kwargs):
+
+        self._default_spectrum_viewer_reference_name = kwargs.get(
+            "spectrum_viewer_reference_name", "spectrum-viewer"
+        )
+
+        super().__init__(*args, **kwargs)
+
+        self.function = SelectPluginComponent(self,
+                                              items='function_items',
+                                              selected='function_selected',
+                                              manual_options=['Mean', 'Min', 'Max', 'Sum'])  # noqa
+
+        self.add_results.viewer.filters = ['is_spectrum_viewer']
+
+    @property
+    def user_api(self):
+        return PluginUserApi(self, expose=('function', 'spatial_subset',
+                                           'add_results', 'collapse_to_spectrum'))
+
+    def collapse_to_spectrum(self, add_data=True, **kwargs):
+        """
+        Collapse over the spectral axis.
+
+        Parameters
+        ----------
+        add_data : bool
+            Whether to load the resulting data back into the application according to
+            ``add_results``.
+        kwargs : dict
+            Additional keyword arguments passed to the NDDataArray collapse operation.
+            Examples include ``propagate_uncertainties`` and ``ignore_masked_data``.
+        """
+        # get glue Data objects for the spectral cube and uncertainties
+        [spectral_cube] = self._app.get_data_from_viewer(
+            self._app._jdaviz_helper._default_flux_viewer_reference_name,
+            include_subsets=False
+        ).values()
```

Comment:
could this cause issues if the user creates a moment map and sends it to be shown in the flux-viewer (as this plugin would then pull a 2d image instead of the flux cube)?

Reply:
Yes that would be problematic here. We wouldn't want the user to think they could extract a spectrum from a moment map... should we raise an error if the flux-viewer has >1 Data?

Reply:
Now reported as a bug [:cat:](https://jira.stsci.edu/browse/JDAT-3790)